x/committee/keeper/proposal.go

```diff
@@ -81,34 +87,143 @@ func (k Keeper) AddVote(ctx sdk.Context, proposalID uint64, voter sdk.AccAddress
-	pr, found := k.GetProposal(ctx, proposalID)
-	if !found {
-		return false, sdkerrors.Wrapf(types.ErrUnknownProposal, "%d", proposalID)
+// ValidatePubProposal checks if a pubproposal is valid.
+func (k Keeper) ValidatePubProposal(ctx sdk.Context, pubProposal types.PubProposal) (returnErr error) {
+	if pubProposal == nil {
+		return sdkerrors.Wrap(types.ErrInvalidPubProposal, "pub proposal cannot be nil")
 	}
-	com, found := k.GetCommittee(ctx, pr.CommitteeID)
-	if !found {
-		return false, sdkerrors.Wrapf(types.ErrUnknownCommittee, "%d", pr.CommitteeID)
+	if err := pubProposal.ValidateBasic(); err != nil {
+		return err
+	}
+
+	if !k.router.HasRoute(pubProposal.ProposalRoute()) {
+		return sdkerrors.Wrapf(types.ErrNoProposalHandlerExists, "%T", pubProposal)
 	}
 
-	numVotes := k.TallyVotes(ctx, proposalID)
+	// Run the proposal's changes through the associated handler using a cached version of state to ensure changes are not permanent.
+	cacheCtx, _ := ctx.CacheContext()
+	handler := k.router.GetRoute(pubProposal.ProposalRoute())
 
-	proposalResult := sdk.NewDec(numVotes).GTE(com.VoteThreshold.MulInt64(int64(len(com.Members))))
+	// Handle an edge case where a param change proposal causes the proposal handler to panic.
+	// A param change proposal with a registered subspace value but unregistered key value will cause a panic in the param change proposal handler.
+	// This defer will catch panics and return a normal error: `recover()` gets the panic value, then the enclosing function's return value is swapped for an error.
+	// reference: https://stackoverflow.com/questions/33167282/how-to-return-a-value-in-a-go-function-that-panics?noredirect=1&lq=1
+	defer func() {
+		if r := recover(); r != nil {
+			returnErr = sdkerrors.Wrapf(types.ErrInvalidPubProposal, "proposal handler panicked: %s", r)
+		}
+	}()
 
-	return proposalResult, nil
+	if err := handler(cacheCtx, pubProposal); err != nil {
+		return err
+	}
+	return nil
 }
 
-// TallyVotes counts all the votes on a proposal
-func (k Keeper) TallyVotes(ctx sdk.Context, proposalID uint64) int64 {
+func (k Keeper) ProcessProposals(ctx sdk.Context) {
+
+	k.IterateProposals(ctx, func(proposal types.Proposal) bool {
+
+		committee, found := k.GetCommittee(ctx, proposal.CommitteeID)
+		if !found {
+			return false // TODO: If committees can't be removed while having active proposals, should panic here
+		}
 
+		if !proposal.HasExpiredBy(ctx.BlockTime()) {
+			if committee.GetTallyOption() == types.FirstPastThePost {
+				passed := k.GetProposalResult(ctx, proposal.ID, committee)
+				if passed {
+					outcome := k.attemptEnactProposal(ctx, proposal)
+					k.CloseProposal(ctx, proposal, outcome)
+				}
+			}
+		} else {
+			passed := k.GetProposalResult(ctx, proposal.ID, committee)
+			outcome := types.Failed
+			if passed {
+				outcome = k.attemptEnactProposal(ctx, proposal)
+			}
+			k.CloseProposal(ctx, proposal, outcome)
+		}
+		return false
```

Comment:
Logic looks good. Test cases I can think of:

1. First pass the post proposal passes before deadline, get's enacted
2. FPTP proposal fails after deadline, ensure that propsal get's deleted
3. Deadline proposal only gets talled at/after deadline
4. Pass a proposal (I think this can be done with a god committee) that fails `attemptEnactProposal`, ensure that proposal is deleted

Reply:
Added all these test cases in https://github.com/Kava-Labs/kava/pull/902/commits/f0e31b16cffc593cc2e9bdef25dac089b03eb9f7. Got hung up for a bit checking marshaled event attributes against `types.ProposalOutcomes`. Probably makes sense to clean up the helper function that does the check, but since it's only used in the test suite I don't think it's a blocker for this merge.

Reply:
There's some helper functions in the sdk for converting events to strings rather than bytes. Might be easier to work with
https://github.com/cosmos/cosmos-sdk/blob/042dad84b3eed262a8a34b21a2c7166fef3e41a6/types/events.go#L208-L216